### PR TITLE
script: Try to conform a bit more to "The event loop processing model"

### PR DESCRIPTION
**Pull request by emilio**
Let's see what breaks.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] There are tests for these changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13864)

### CODE REVIEW

**Comment by KiChjang** [comment]
@bors-servo retry
* #13887

**Comment by emilio** [comment]
That's https://github.com/servo/servo/issues/13887.

**Comment by emilio** [comment]
I suspect #13887 might become perma-failure due to the proper event ordering, and https://github.com/servo/servo/issues/13865.

In any case, it's the same underlying issue, and I think it'd be acceptable to mark them as failing since they were passing before due to a huge hack we had.

**Comment by emilio** [comment]
@bors-servo: try

Fixed the timeout, which after investigation turned out to be an issue with the merge with the patch in #13866.

**Comment by emilio** [comment]
@bors-servo retry

I think the last test failure was intermittent too.

**Comment by emilio** [comment]
Review ping @pcwalton? (Not special rush, this is not really prone to bitrot)

**Comment by emilio** [comment]
I'm wondering (it can happen with and without this PR though) what happens if:

 * A resize event is triggered.
 * That resize event callback removes a DOM node for which a transition event has been triggered.
 * GC happens.
 * We process the transition event, converting the `UnsafeNode` back to a node.

I'm wary since this comes from #13056 (where I warned about the same thing). @KiChjang do you know if this can happen?

We can fix that pre-processing the transition event nodes, which is probably ok, but worth thinking if it's necessary.

File: components/script_traits/lib.rs [hunk 1]
```diff
@@ -113,8 +113,8 @@ pub enum LayoutControlMsg {
     ExitNow,
     /// Requests the current epoch (layout counter) from this layout.
     GetCurrentEpoch(IpcSender<Epoch>),
-    /// Asks layout to run another step in its animation.
-    TickAnimations,
+    /// Asks layout to run another step in its animations.
+    TickAnimations(bool /* script_animation_callbacks_present */),
```
Comment (pcwalton) [inline components/script_traits/lib.rs]: I kind of wonder whether these should be struct-like enum variants.

**Comment by emilio** [comment]
On Tue, Nov 08, 2016 at 09:03:26AM -0800, Patrick Walton wrote:
> I kind of wonder whether these should be struct-like enum variants.

Agreed, most of our Message enums in general IMO.

I'll convert this one when I get back to this PR, and open issues for
the others.

**Comment by jdm** [comment]
What is the plan for this PR? It's not clear what the next step is for it.

**Comment by pcwalton** [comment]
This seems fine to me.

**Comment by jdm** [comment]
The concern in https://github.com/servo/servo/pull/13864#issuecomment-257833248 sounds exactly like #14972 which has been observed in the field.

**Comment by emilio** [comment]
@jdm that concern is pre-existing though, so I'll try to rebase this on top master. Addressing that shouldn't (hopefully) be extremely hard.

**Comment by nox** [comment]
Ok, this is what I suspected. This PR isn't even showing itself in Homu's queue at all.

**Comment by nox** [comment]
```
diff --git a/tests/wpt/mozilla/meta/MANIFEST.json b/tests/wpt/mozilla/meta/MANIFEST.json
index 5d7438c..46c69ce 100644
--- a/tests/wpt/mozilla/meta/MANIFEST.json
+++ b/tests/wpt/mozilla/meta/MANIFEST.json
@@ -20010,7 +20010,7 @@
    "testharness"
   ],
   "css/animations/basic-transition.html": [
-   "6f4319854de92d7aee5ef425e17e80f9c39675df",
+   "4c1d67d8c180880c7e11c95979ef0f0346e8ce31",
    "testharness"
   ],
   "css/animations/mixed-units.html": [
```

**Comment by nox** [comment]
What's the status on this?

**Comment by jdm** [comment]
#14972 is being solved by #16295. We should be sure that this PR does not introduce similar problems.

**Comment by emilio** [comment]
Yes, this shouldn't affect it. Thanks for fixing that @jdm!

**Comment by nox** [comment]
Superseded by #18611.

**Pull request closed without merging**